// @google/events/firebase/remoteconfig/v1/RemoteConfigEventData.js
"use strict";
/**
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.toRemoteConfigEventData = exports.UpdateTypeEnum = exports.UpdateOriginEnum = void 0;
var UpdateOriginEnum;
(function (UpdateOriginEnum) {
    UpdateOriginEnum["AdminSDKNode"] = "ADMIN_SDK_NODE";
    UpdateOriginEnum["Console"] = "CONSOLE";
    UpdateOriginEnum["RESTAPI"] = "REST_API";
    UpdateOriginEnum["RemoteConfigUpdateOriginUnspecified"] = "REMOTE_CONFIG_UPDATE_ORIGIN_UNSPECIFIED";
})(UpdateOriginEnum = exports.UpdateOriginEnum || (exports.UpdateOriginEnum = {}));
var UpdateTypeEnum;
(function (UpdateTypeEnum) {
    UpdateTypeEnum["ForcedUpdate"] = "FORCED_UPDATE";
    UpdateTypeEnum["IncrementalUpdate"] = "INCREMENTAL_UPDATE";
    UpdateTypeEnum["RemoteConfigUpdateTypeUnspecified"] = "REMOTE_CONFIG_UPDATE_TYPE_UNSPECIFIED";
    UpdateTypeEnum["Rollback"] = "ROLLBACK";
})(UpdateTypeEnum = exports.UpdateTypeEnum || (exports.UpdateTypeEnum = {}));
/**
 * Cast a raw JSON object to a typed event (useful for IDE autocompletion).
 * @param {object} json The JSON object
 * @return {RemoteConfigEventData} The object with type annotations
 */
const toRemoteConfigEventData = (json) => {
    return json;
};
exports.toRemoteConfigEventData = toRemoteConfigEventData;
